### PR TITLE
JSON backend

### PR DESCRIPTION
**Pull request by franzpoeschel**
Implementation of a JSON backend for openPMD.

Todo/Issues:
* [x] user documentation
* [x] Some of the methods in `AbstractIOHandlerImpl.hpp` did not behave exactly as documented (modulo misunderstandings on my side). I will address this separately.
* [ ] see the discussion in #65 
* [x] integration with the build system: #384
* [x] remove commits from #337 and rebase onto master
* [x] put appropriate license, copyright and author headers in every file

### CODE REVIEW

**Comment by ax3l** [comment]
@franzpoeschel I know the changes were quite heavy recently (big sorry again). Can you please try the following to update this branch? I would like to remove the number of commits and would like to have no merge commit in it.

```bash
# backup in case things go south
git checkout -b json-backend-bak
# switch back to json-backend 
git checkout -

# rebase: squash
git fetch --all
# I assume your "git remote -v" for this mainline repo is called "mainline"
git rebase -i mainline
# an editor will open
#   change "pick" to "s" ("squash") to all your commits *but the first entry*
#   remove the lines with "merge commits",
#   remove commits a7a7974 and 176d37d 
# save
# solve merge conflicts one

# if anything goes very wrong
# git rebase --abort

# in the end, update the PR via
git push -f
```

**Comment by franzpoeschel** [comment]
Concerning user documentation, is there a place where the file formats are / should be documented?

@ax3l Done, thanks for the steps.

**Comment by ax3l** [comment]
>Concerning user documentation, is there a place where the file formats
>are / should be documented?

Yes, it's even recommended :) You can update the documentation files (.rst) in the same PR that lands the feature :)

**Comment by ax3l** [comment]
@franzpoeschel I added complete build system support for you in #384.
Is it ok if I rebase this branch and force-push update it against those changes?

**Comment by ax3l** [comment]
Pushed the rebased branch :)
Please make sure that you don't have uncommited/unpushed *local* changes before continuing to work on this. Fetch the latest online version, revert the last local commit if they were identical before my rebase and pull the fresh commit.

```bash
git status -sb
# nothing uncommited?
# latest commits were pushed online?

# assuming yes

# fetch the updates I pushed on your branch and remove local "outdated" commit
git fetch --all
git reset --hard HEAD~
git pull --ff-only
```

After that, you can continue as usual.

If you *do* have local changes that were not yet pushed - commit them and rebase the branch against the online commit (also removing the one outdated commit but keeping all other changes on top).

```bash
git fetch --all

git rebase -i mainline/dev
# remove outdated commit
# keep all others
```

**Comment by franzpoeschel** [comment]
Thanks for the help and the detailed instructions!

File: include/openPMD/auxiliary/Memory.hpp
Comment (ax3l) [inline include/openPMD/auxiliary/Memory.hpp]: little formatting issue :)

File: include/openPMD/auxiliary/Filesystem.hpp
Comment (ax3l) [inline include/openPMD/auxiliary/Filesystem.hpp]: can this  dead code be removed? :)

File: include/openPMD/IO/JSON/JSONFilePosition.hpp
Comment (ax3l) [inline include/openPMD/IO/JSON/JSONFilePosition.hpp:39]: can you please implement the bidy in a `.cpp` file? Avoids pulling in the public `nlohmann/json.hpp` header into our public API.

File: include/openPMD/Datatype.hpp
Comment (ax3l) [inline include/openPMD/Datatype.hpp]: maybe no empy newline here. You can put the `OPENPMD_TEMPLATE_OPERATOR` define block above the doxygen block

File: docs/source/backends/json_example.json
Comment (ax3l) [inline docs/source/backends/json_example.json:1]: When we are finished with the review (until then, pls update here):
Shall we rather put this file into https://github.com/openPMD/openPMD-example-datasets ?

File: docs/source/backends/json.rst
Comment (ax3l) [inline docs/source/backends/json.rst:82]: maybe link to https://github.com/openPMD/openPMD-example-datasets when it's included there?

File: docs/source/backends/json_example.json
Comment (ax3l) [inline docs/source/backends/json_example.json]: why do you add a `subgroups` directory here and before `data`? Can we skip those maybe?

File: test/SerialIOTest.cpp
Comment (ax3l) [inline test/SerialIOTest.cpp]: Do you think we can just generalize the `hdf5_`/`adios_`/`json_` write test into a helper function and loop with a `Series` object for each backend? Otherwise we copy so much code...

**Review by ax3l — COMMENTED** [review]
Dang, I reviewed this a week ago and forgot to submit my comments.

File: docs/source/backends/json_example.json
Comment (franzpoeschel) [reply docs/source/backends/json_example.json]: Short answer: separating namespaces. If this indirection were dropped, the subgroups' and datasets' names will have the same namespace as all other JSON objects at that level, including the objects `attributes` and for the top-level group `platform_byte_widths` (meaning you can't have groups with those names). This seems like a nasty corner case for a user of the API to run into, hence I chose clarity at the expense of a certain informational overhead as is explained in [here](https://github.com/openPMD/openPMD-api/blob/ab2fb040060089a72acaa2c14a8c72a909df2b66/docs/source/backends/json.rst).
Do you have a better solution? I can imagine wrapping the subgroups into quotes or maybe reserving a prefix for "keywords" (such as `__attributes` maybe). Would this be preferrable?

File: include/openPMD/auxiliary/Filesystem.hpp
Comment (franzpoeschel) [reply include/openPMD/auxiliary/Filesystem.hpp]: done ;)

File: include/openPMD/IO/JSON/JSONFilePosition.hpp
Comment (franzpoeschel) [reply include/openPMD/IO/JSON/JSONFilePosition.hpp:39]: Done.... but I still need to include `nlohmann/json.hpp` in the header file?
This being said, `JSONFilePosition.hpp` is only ever included in `JSONIOHandlerImpl.hpp`, which is included in `JSONIOHandler.hpp` and that header file is only included in `*.cpp` files, so it should be fine?

File: include/openPMD/auxiliary/Memory.hpp
Comment (franzpoeschel) [reply include/openPMD/auxiliary/Memory.hpp]: Ah, good you spotted that, those includes were left from when I added a function to that file which I later removed again. Removed the includes.

File: include/openPMD/Datatype.hpp
Comment (franzpoeschel) [reply include/openPMD/Datatype.hpp]: done

File: docs/source/backends/json_example.json
Comment (ax3l) [reply docs/source/backends/json_example.json]: Yes, there is a risk of collisions which in practice should not be too often the case. Nevertheless, rather adding a *dict of attributes* inside an `attributes` group would probably be best (no `__`). This keeps the collisions minimal and at the same time is very readable. But that's already the case, no?

The same is true for data sets and groups. Without the `subgroups` "namespace" one can not write the same group and data set name in the same level and neither can be named "attributes". But imho that's fine.

File: include/openPMD/IO/JSON/JSONFilePosition.hpp
Comment (ax3l) [reply include/openPMD/IO/JSON/JSONFilePosition.hpp:39]: No, if no symbols/declarations are used from `nlohmann/json.hpp` in the header file, then do not include it there.

> This being said, [...]

That sounds good, I wasn't sure.

File: test/SerialIOTest.cpp
Comment (franzpoeschel) [reply test/SerialIOTest.cpp]: Pushed a commit that does this. This leaves only the tests using samples as backend-specific tests.

File: docs/source/backends/json_example.json
Comment (franzpoeschel) [reply docs/source/backends/json_example.json]: >But that's already the case, no?

yep

I changed it now, so attributes keep their separate namespace but datasets and subgroups are now directly linked into the containing group's object, see [here](https://github.com/franzpoeschel/openPMD-api/blob/json-backend/docs/source/backends/json_example.json).

File: include/openPMD/IO/JSON/JSONFilePosition.hpp
Comment (franzpoeschel) [reply include/openPMD/IO/JSON/JSONFilePosition.hpp:39]: >if no symbols/declarations are used from nlohmann/json.hpp in the header file

They are. `JSONFilePosition.hpp` is implemented as a wrapper around `json::json_pointer`.

File: docs/source/backends/json_example.json
Comment (ax3l) [reply docs/source/backends/json_example.json]: Looks good!
Do you think you need the `"object_type": "dataset"` and `"object_type": "group"` keys? They don't hurt, but I wonder if one can already guess by the name (e.g. `data`) and content already.

Comment (franzpoeschel) [reply docs/source/backends/json_example.json]: I added those for the `listPaths` and `listDatasets` operations. I did not want to rely on checking the structure of JSON objects to distinguish groups from datasets, so I made the distinction explicit.
I mean, both layouts (i.e. of groups vs. datasets) should actually be computationally distinguishable, but should we go that length?

Comment (ax3l) [reply docs/source/backends/json_example.json]: I get the first part, but think it's probably ok to check if a `data` members is present.

Can please you elaborate what you mean with the last sentence (`computationally distinguishable`)?

Comment (franzpoeschel) [reply docs/source/backends/json_example.json]: >Can please you elaborate what you mean with the last sentence (computationally distinguishable)?

Pretty much what you said already, datasets and groups can be distinguished by inspecting their members.
>think it's probably ok to check if a data members is present

Probably not a good idea if you look at [line 28](https://github.com/franzpoeschel/openPMD-api/blob/json-backend/docs/source/backends/json_example.json) of the example JSON file ;)
The safest bet would probably be to check whether the `extent` key exists *and* whether it points to a JSON array. That should work, but it *feels* like a hack. Should I do it this way?

Comment (ax3l) [reply docs/source/backends/json_example.json]: Why do we need `extent`? Can't we determine this from a list (of lists (of lists)) by querying their lengths, even in ND-cases?

Would checking for "contains a data member that contains a list" be enough to detect a low-level "data set" vs. low-level "group"?

I just want to avoid introducing too many non-standard attributes/sub-groups/naming conventions. It might be that we need some, since we don't have as in XML/HDF5/ADIOS/... attributes on a path that are distinct from attributes on data sets. But we should keep it minimal, e.g. for attributes it's reasonable to have this sub-group and also to have a "data" key.

Comment (franzpoeschel) [reply docs/source/backends/json_example.json]: >Why do we need extent? Can't we determine this from a list (of lists (of lists)) by querying their lengths, even in ND-cases?

In principle, yes. It would just require pre-filling the dataset with `null` values to be sure that it does actually have the specified dimensions. (Currently, the data array is just as big as necessary for holding the chunks that have actually been written).
Would you prefer this?

Comment (franzpoeschel) [reply docs/source/backends/json_example.json]: The latest commit eliminates the `extent` field, next up will be the elimination of `object_type`.

Comment (ax3l) [reply docs/source/backends/json_example.json]: > pre-filling the dataset with null values

makes probably sense, this only works in serial anyway, right?

Comment (ax3l) [reply docs/source/backends/json_example.json]: > The latest commit eliminates the extent field, next up will be the elimination of object_type.

Can you please update the checked-in example file? :)

Comment (franzpoeschel) [reply docs/source/backends/json_example.json]: >this only works in serial anyway, right?

right
`object_type` is no more and the example file and documentation are updated.

Comment (ax3l) [reply docs/source/backends/json_example.json]: cool, thanks! looks good :)

**Comment by ax3l** [comment]
@franzpoeschel I think I can implement #389 now :)
That means: when it is merged you can remove the changes you made to the vendored `nlohmann-json` files, which is a good thing. If you like, rebase and squash some commits when you are on it? (But again, wait for both until #389 is merged) :)

**Comment by ax3l** [comment]
@franzpoeschel ready for rebase, remove of the JSON third party patch and some squashing of commits :)

**Comment by ax3l** [comment]
@franzpoeschel just do (in case your `git remote` for the main repo is called `mainline`, see `git remote -v`):
```bash
git fetch mainline
git rebase -i mainline/dev
git push -f
```

**Comment by ax3l** [comment]
Ouch, looks like #389 did not trigger `-isystem` yet... (Sorry!)

**Comment by ax3l** [comment]
@franzpoeschel I don't know why this did not work, but you can fix it by adding an extra commit changing the following:
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 2c73e2d..418428f 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -127,8 +127,10 @@ else()
 endif()
 if(openPMD_HAVE_JSON)
     add_library(openPMD::thirdparty::nlohmann_json INTERFACE IMPORTED)
-    target_link_libraries(openPMD::thirdparty::nlohmann_json
-        INTERFACE nlohmann_json::nlohmann_json)
+    #target_link_libraries(openPMD::thirdparty::nlohmann_json
+    #    INTERFACE nlohmann_json::nlohmann_json)
+    get_target_property(lib_include_dirs nlohmann_json::nlohmann_json INTERFACE_INCLUDE_DIRECTORIES)
+    target_include_directories(openPMD::thirdparty::nlohmann_json SYSTEM INTERFACE ${lib_include_dirs})
 endif()
 
 # external library: HDF5 (optional)
```

**Comment by ax3l** [comment]
I pushed it for you.

**Review by ax3l — APPROVED** [review]
Hurray, thanks a lot this is great! :sparkles:

**Pull request merged**
